erpc/networks_test.go

```diff
@@ -5844,6 +5844,124 @@ func TestNetwork_InFlightRequests(t *testing.T) {
 	})
 }
 
+func TestNetwork_SkippingUpstreams(t *testing.T) {
+	t.Run("NotSkippedRecentBlockNumberForFullNodeUpstream", func(t *testing.T) {
+		util.ResetGock()
+		defer util.ResetGock()
+
+		ctx, cancel := context.WithCancel(context.Background())
+		defer cancel()
+		network := setupTestNetworkWithFullNodeUpstream(t, ctx, nil, nil)
+
+		requestBytes := []byte(`{"jsonrpc":"2.0","method":"eth_getBalance","params":["0x0000000000000000000000000000000000000000", "0x1273c18"]}`)
+		gock.New("http://rpc1.localhost").
+			Post("/").
+			Persist().
+			Filter(func(request *http.Request) bool {
+				return strings.Contains(util.SafeReadBody(request), "eth_getBalance")
+			}).
+			Reply(200).
+			BodyString(`{"jsonrpc":"2.0","id":1,"result":"0x0"}`)
+		req := common.NewNormalizedRequest(requestBytes)
+		resp, err := network.Forward(context.Background(), req)
+		assert.NoError(t, err)
+		assert.NotNil(t, resp)
+	})
+
+	t.Run("SkippedHistoricalBlockNumberForFullNodeUpstream", func(t *testing.T) {
+		util.ResetGock()
+		defer util.ResetGock()
+
+		ctx, cancel := context.WithCancel(context.Background())
+		defer cancel()
+		network := setupTestNetworkWithFullNodeUpstream(t, ctx, nil, nil)
+
+		requestBytes := []byte(`{"jsonrpc":"2.0","method":"eth_getBalance","params":["0x0000000000000000000000000000000000000000", "0x1"]}`)
+		gock.New("http://rpc1.localhost").
+			Post("/").
+			Persist().
+			Filter(func(request *http.Request) bool {
+				return strings.Contains(util.SafeReadBody(request), "eth_getBalance")
+			}).
+			Reply(500).
+			Delay(1 * time.Second).
+			BodyString(`{"jsonrpc":"2.0","id":1,"error":{"code":-32014,"message":"Internal error"}}`)
+		req := common.NewNormalizedRequest(requestBytes)
+		resp, err := network.Forward(context.Background(), req)
+		assert.Error(t, err)
+		assert.Nil(t, resp)
+	})
+
+	t.Run("NotSkippedHistoricalBlockForArchiveNodeUpstream", func(t *testing.T) {
+		util.ResetGock()
+		defer util.ResetGock()
+
+		ctx, cancel := context.WithCancel(context.Background())
+		defer cancel()
+		network := setupTestNetworkWithArchiveNodeUpstream(t, ctx, nil, nil)
+
+		requestBytes := []byte(`{"jsonrpc":"2.0","method":"eth_getBalance","params":["0x0000000000000000000000000000000000000000", "0x1"]}`)
+		gock.New("http://rpc1.localhost").
+			Post("/").
+			Persist().
+			Filter(func(request *http.Request) bool {
+				return strings.Contains(util.SafeReadBody(request), "eth_getBalance")
+			}).
+			Reply(200).
+			BodyString(`{"jsonrpc":"2.0","id":1,"result":"0x0"}`)
+		req := common.NewNormalizedRequest(requestBytes)
+		resp, err := network.Forward(context.Background(), req)
```

Comment:
```suggestion
		resp, err := network.Forward(ctx, req)
```
So that when test is finished request is also forced to cancel (without it we sometimes have gock race conditions)